### PR TITLE
Python 3 updates for SFF support and tests

### PR DESCRIPTION
**Pull request by peterjc**
This pull request addresses in part issue #2490.

See also discussion on #2496.

- [X] I hereby agree to dual licence this and any previous contributions under both
the _Biopython License Agreement_ **AND** the _BSD 3-Clause License_.

- [X] I have read the ``CONTRIBUTING.rst`` file, have run ``flake8`` locally, and
understand that AppVeyor and TravisCI will be used to confirm the Biopython unit
tests and style checks pass with these changes.

- [X] I have added my name to the alphabetical contributors listings in the files
``NEWS.rst`` and ``CONTRIB.rst`` as part of this pull request, am listed
already, or do not wish to be listed. (*This acknowledgement is optional.*)

### CODE REVIEW

File: Bio/SeqIO/SffIO.py [hunk 1]
```diff
@@ -317,6 +291,8 @@ def _sff_file_header(handle):
     data = handle.read(31)
     if not data:
         raise ValueError("Empty file.")
+    elif isinstance(data, str):
```
Comment (mdehoon) [inline Bio/SeqIO/SffIO.py]: The pythonic way would be to call `struct.unpack(fmt, data)` below in a `try: except:` block. `struct.unpack` will raise a `TypeError` if `data` is not a bytes-like object. Then you don't need to use `isinstance` here.

File: Tests/test_SeqIO_index.py [hunk 2]
```diff
@@ -46,15 +45,11 @@ def add_prefix(key):
...
-    # insists on giving byte strings (i.e. binary mode)
-    # See http://bugs.python.org/issue13989
-    if sys.version_info[0] < 3 or format in SeqIO._BinaryFormats:
-        return gzip.open(filename)
-    handle = gzip.open(filename)
-    data = handle.read()  # bytes!
-    handle.close()
-    return StringIO(data.decode())
+    """Open gzip file in either binary or text mode as needed."""
+    if format in SeqIO._BinaryFormats:
```
Comment (mdehoon) [inline Tests/test_SeqIO_index.py]: Some day we should try and get rid of `SeqIO._BinaryFormats`

File: Bio/SeqIO/SffIO.py [hunk 3]
```diff
@@ -1159,14 +1135,23 @@ def __init__(self, handle, index=True, xml=None):
...
+        try:
+            # This should fail...
+            handle.write("ERROR - text mode handle\n")
+            raise ValueError(
+                "SFF files must NOT be opened in text mode, binary required."
+            )
+        except TypeError:
+            # Good, we wanted a binary handle:
+            # e.g. TypeError: a bytes-like object is required, not 'str'
+            pass
```
Comment (mdehoon) [inline Bio/SeqIO/SffIO.py]: Actually here you could use `handle.write(b"")` to explicitly test if the handle was opened in binary mode.

File: Tests/test_SeqIO_index.py (hunk 2 above)
Comment (peterjc) [reply Tests/test_SeqIO_index.py]: Yes, and this pull request is a small step in that direction.

File: Bio/SeqIO/SffIO.py (hunk 3 above)
Comment (peterjc) [reply Bio/SeqIO/SffIO.py]: I didn't think of that - almost but not quite a NOP (no operation), that does seem more elegant.

File: Bio/SeqIO/SffIO.py (hunk 1 above)
Comment (peterjc) [reply Bio/SeqIO/SffIO.py]: OK, that makes sense.

Comment (peterjc) [reply Bio/SeqIO/SffIO.py]: Agreed.

**Comment by peterjc** [comment]
Thank you - your input definitely improved this.

**Pull request merged**